test/test_common/utility.cc

```diff
@@ -139,32 +139,60 @@ Stats::TextReadoutSharedPtr TestUtility::findTextReadout(Stats::Store& store,
   return findByName(store.textReadouts(), name);
 }
 
-void TestUtility::waitForCounterEq(Stats::Store& store, const std::string& name, uint64_t value,
-                                   Event::TestTimeSystem& time_system) {
+AssertionResult TestUtility::waitForCounterEq(Stats::Store& store, const std::string& name,
+                                              uint64_t value, Event::TestTimeSystem& time_system,
+                                              std::chrono::milliseconds timeout) {
+  auto end_time = time_system.monotonicTime() + timeout;
   while (findCounter(store, name) == nullptr || findCounter(store, name)->value() != value) {
     time_system.advanceTimeWait(std::chrono::milliseconds(10));
+    if (time_system.monotonicTime() >= end_time) {
+      ENVOY_LOG_MISC(trace, "Timed out waiting for {} to be {}", name, value);
+      return AssertionFailure() << fmt::format("timed out waiting for {} to be {}", name, value);
+    }
   }
+  return AssertionSuccess();
 }
 
-void TestUtility::waitForCounterGe(Stats::Store& store, const std::string& name, uint64_t value,
-                                   Event::TestTimeSystem& time_system) {
+AssertionResult TestUtility::waitForCounterGe(Stats::Store& store, const std::string& name,
+                                              uint64_t value, Event::TestTimeSystem& time_system,
+                                              std::chrono::milliseconds timeout) {
+  auto end_time = time_system.monotonicTime() + timeout;
   while (findCounter(store, name) == nullptr || findCounter(store, name)->value() < value) {
     time_system.advanceTimeWait(std::chrono::milliseconds(10));
+    if (time_system.monotonicTime() >= end_time) {
+      ENVOY_LOG_MISC(trace, "Timed out waiting for {} to be {}", name, value);
+      return AssertionFailure() << fmt::format("timed out waiting for {} to be {}", name, value);
+    }
   }
+  return AssertionSuccess();
 }
 
-void TestUtility::waitForGaugeGe(Stats::Store& store, const std::string& name, uint64_t value,
-                                 Event::TestTimeSystem& time_system) {
+AssertionResult TestUtility::waitForGaugeGe(Stats::Store& store, const std::string& name,
+                                            uint64_t value, Event::TestTimeSystem& time_system,
+                                            std::chrono::milliseconds timeout) {
+  auto end_time = time_system.monotonicTime() + timeout;
   while (findGauge(store, name) == nullptr || findGauge(store, name)->value() < value) {
     time_system.advanceTimeWait(std::chrono::milliseconds(10));
+    if (time_system.monotonicTime() >= end_time) {
+      ENVOY_LOG_MISC(trace, "Timed out waiting for {} to be {}", name, value);
```

Comment:
I think this will be the case, but if you hit an assertion failure, i think it will log anyway. Maybe you can test this out be running ads_integration_test and spuriously increasing the value?

Reply:
yep, it logs; removed